src/inmanta/agent/handler.py

```diff
@@ -419,7 +419,52 @@ def _log_msg(self, level: int, msg: str, *args: object, exc_info: bool = False,
 
 
 @stable_api
-class ResourceHandler(object):
+class HandlerABC(ABC):
+    """
+    Top-level abstract base class all handlers should inherit from.
+    """
+
+    @abstractmethod
+    def pre(self, ctx: HandlerContext, resource: resources.Resource) -> None:
+        """
+        Method executed before a handler operation (Facts, dryrun, real deployment, ...) is executed. Override this method
+        to run before an operation.
+
+        :param ctx: Context object to report changes and logs to the agent and server.
+        :param resource: The resource to query facts for.
+        """
+
+    @abstractmethod
```

Comment:
I don't think this one or `pre` should be marked as `abstractmethod` because that requires concrete classes to override them. We want their default behavior to  be a NOOP, don't we?

Reply:
I agree, that's the route I took on my PR implementing the `DiscoveryHandler`, I'll change it here as well